include/nanobind/ndarray.h

```diff
@@ -105,261 +141,125 @@ template <size_t... S> struct ndim_shape<std::index_sequence<S...>> {
-};
-
-template <> struct ndarray_arg<any_contig> {
-    static constexpr size_t size = 0;
-    static constexpr auto name = const_name("order='*'");
-    static void apply(ndarray_req &tr) { tr.req_order = '\0'; }
+/// Sentinel type to initialize ndarray_config_t<>
+struct unused {
+    using type = void;
+    static constexpr int value = 0;
+    static constexpr auto name = descr<0>();
 };
 
-template <typename T> struct ndarray_arg<T, enable_if_t<T::is_device>> {
-    static constexpr size_t size = 0;
-    static constexpr auto name = const_name("device='") + T::name + const_name('\'');
-    static void apply(ndarray_req &tr) { tr.req_device = (uint8_t) T::value; }
-};
+/// ndarray_config describes a requested array configuration
+struct ndarray_config {
+    int device_type;
+    char order;
+    bool ro;
+    dlpack::dtype dtype;
+    int32_t ndim;
+    int64_t *shape;
 
-template <typename... Ts> struct ndarray_info {
-    using scalar_type = void;
-    using shape_type = void;
-    constexpr static bool ReadOnly = false;
-    constexpr static auto name = const_name("ndarray");
-    constexpr static ndarray_framework framework = ndarray_framework::none;
-    constexpr static char order = '\0';
+    template <typename T> ndarray_config(T)
+        : device_type(T::DeviceType::value),
+          order((char) T::Order::value),
+          ro(std::is_const_v<typename T::Scalar>),
+          dtype(nanobind::dtype<typename T::Scalar>()),
+          ndim(T::N),
+          shape(nullptr) { }
 };
 
-template <typename T, typename... Ts> struct ndarray_info<T, Ts...>  : ndarray_info<Ts...> {
-    using scalar_type =
-        std::conditional_t<
-                detail::is_ndarray_scalar_v<T> &&
-                    std::is_void_v<typename ndarray_info<Ts...>::scalar_type>,
-                T, typename ndarray_info<Ts...>::scalar_type>;
-
-    constexpr static bool ReadOnly = ndarray_info<Ts...>::ReadOnly ||
-            (detail::is_ndarray_scalar_v<T> && std::is_const_v<T>);
-};
+/// ndarray_config_t collects nd-array template parameters in a structured way.
+/// Its "storage" is purely based on types members
+template <typename /* SFINAE */ = int, typename...> struct ndarray_config_t;
 
-template <typename... Ts> struct ndarray_info<ro, Ts...> : ndarray_info<Ts...> {
-    constexpr static bool ReadOnly = true;
+template <> struct ndarray_config_t<int> {
+    using Framework = no_framework;
+    using Scalar = void;
+    using Shape = unused;
+    using Order = unused;
+    using DeviceType = unused;
+    static constexpr int32_t N = -1;
 };
 
-template <ssize_t... Is, typename... Ts> struct ndarray_info<shape<Is...>, Ts...> : ndarray_info<Ts...> {
-    using shape_type =
-        std::conditional_t<
-                std::is_void_v<typename ndarray_info<Ts...>::shape_type>,
-                shape<Is...>, typename ndarray_info<Ts...>::shape_type>;
+// Template infrastructure to collect ndarray annotations and fail if duplicates are found
```

Comment:
Does this still fail if duplicates are found? It doesn't appear so

Reply:
No, I removed this part of the code. It seemed too much complexity for too little benefit.